Support cast projection pushdown in oracle

krvikash opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
This PR adds support for cast projection pushdown for the Oracle connector. This uses the framework introduced in https://github.com/trinodb/trino/pull/22203 for cast projection pushdown.
Currently, this PR only adds support for the `CHAR` and `VARCHAR` cast pushdown, which has the following limitations:

1. Pushdown is not supported with `Char` having length > 500
2. Pushdown is not supported with `Varchar` having length > 1000
3. Pushdown is not supported with unbounded varchar
4. Pushdown is not supported with `Char` cast to bounded `Varchar`

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Oracle
* Add support for cast project pushdown with Char and Varchar data type. ({issue}`22728`

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,108 @@
+            if (varcharType.isUnbounded() || varcharType.getBoundedLength() > ORACLE_VARCHAR2_MAX_CHARS) {
+                return Optional.empty();
+            }
+            return Optional.of(new JdbcTypeHandle(OracleTypes.VARCHAR, Optional.of("VARCHAR2"), varcharType.getLength(), Optional.empty(), Optional.empty(), Optional.empty()));
```
"VARCHAR2" 
can it be obtained as for chars
`varcharType.getBaseName()`

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+                new CastTestCase("char(10)", "varchar(1001)"),
+                new CastTestCase("char(10)", "varchar"),
+                new CastTestCase("varchar(10)", "varchar(1001)"),
+                new CastTestCase("varchar(10)", "varchar"),
```
It's pity. This is very common usecase.

ssheikin replied:
can't we do something for varchars?

krvikash replied:
I think we can use `TO_NCLOB` to cast in `varchar` or `varchar(... >1000 CHAR)`. https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/TO_NCLOB.html

Trying it out.

krvikash replied:
Apart from `CAST` pushdown, comparison with `clob` type was not supported for pushdown. So I have made the change to support  `clob` type comparison pushdown.

ssheikin replied:
magic!

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+                new CastTestCase("char(10)", "char(20)"),
+                new CastTestCase("char(10)", "varchar(20)"),
+                new CastTestCase("char(501)", "varchar(20)"),
+                new CastTestCase("char(501)", "varchar"), // No cast will be used during execution, as both the type will be converted to nclob
```
but it's still pushed down. so it's good.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
     }
 
+    @Test
+    public void testCastProjectionPushdown()
```
extract it as a `TestCast` class and inject to `BaseOracleConnectorTest` as `@Nested`

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+    }
+
+    @Test
+    public void testImplicitCastJoinPushdown()
```
`testJoinPushdownWithImplicitCast`
Leave it in this class.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+    public void testExplicitCastJoinPushdown()
+    {
+        String leftTable = "left_table_" + randomNameSuffix();
+        computeActual("CREATE TABLE %s (id int, varchar_50 varchar(50))".formatted(leftTable));
```
`new TestTable`

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+    public void testImplicitCastJoinPushdown()
+    {
+        String leftTable = "left_table_" + randomNameSuffix();
+        computeActual("CREATE TABLE %s (id int, varchar_50 varchar(50))".formatted(leftTable));
```
`new TestTable`

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "LEFT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, 2));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "RIGHT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, null));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "INNER JOIN", "l.varchar_50 = r.varchar_100", List.of(1));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "INNER JOIN", "r.varchar_100 = l.varchar_50", List.of(1));
```
why join criteria in this assertion differs from others?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+    }
+
+    @Test
+    public void testExplicitCastJoinPushdown()
```
I think it can be extracted to `TestCast` class and inject to `BaseOracleConnectorTest` as `@Nested`

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,108 @@
+                targetTypeHandle.get()));
+    }
+
+    private static Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type)
```
This method is the meat of this PR. 
And this is a main opportunity to improve for function pushdown.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -500,4 +617,13 @@ protected String getUser()
+
+    private record CastTestCase(String fromType, String castType)
+    {
+        private CastTestCase
```
please check that `fromType != castType`
https://github.com/trinodb/trino/pull/21939/commits/edc25bf5d855ec6b82cac562d2a5407570446db7#diff-01f7e5fb80c1ce75cf0ea5d3f1933156eb7984967b99b92090bf931e93569696R1274-R1280

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+                new CastTestCase("varchar(1001)", "varchar"), // No cast will be used during execution, as both the type will be converted to nclob
+                new CastTestCase("varchar", "varchar(20)"))) {
+            try (TestTable table = new TestTable(getQueryRunner()::execute, "test_cast_", "(a %s)".formatted(testCase.fromType()))) {
+                assertThat(query("SELECT CAST(a AS %s) FROM %s".formatted(testCase.castType(), table.getName())))
+                        .isFullyPushedDown();
```
please add test for nested CAST:
`CAST(CAST(a AS %s) AS %s)`
https://github.com/trinodb/trino/pull/21939/commits/edc25bf5d855ec6b82cac562d2a5407570446db7#diff-01f7e5fb80c1ce75cf0ea5d3f1933156eb7984967b99b92090bf931e93569696R1055

ssheikin reviewed (commented):
lgtm

wendigo commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -191,6 +195,119 @@ public void testTimestampOutOfPrecisionRounded()
+
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "LEFT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, 2));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "RIGHT JOIN", "l.varchar_50 = r.varchar_100", Arrays.asList(1, null));
+        assertJoinFullyPushedDown(session, leftTable, rightTable, "INNER JOIN", "l.varchar_50 = r.varchar_100", List.of(1));
```
woah

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+                    new CastTestCase("integer", "bigint"),
+                    new CastTestCase("integer", "real"),
+                    new CastTestCase("integer", "double"),
+                    new CastTestCase("integer", "decimal(15)"),
+                    new CastTestCase("integer", "decimal(10, 2)"),
+                    new CastTestCase("integer", "decimal(30, 2)"),
+                    new CastTestCase("smallint", "integer"),
```
Despite the complexity of the implementation 
https://github.com/trinodb/trino/pull/22728/commits/f4204f3c75cd0510e91e22e1dc8f625eb4229744#r1683242993
I guess these return types can be handled too. Could you please try?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+    }
+
+    @Nested
+    class TestCastPushdown
```
please move to a separate file

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteStringComparison.java:
```diff
@@ -70,19 +70,48 @@ public Optional<ParameterizedExpression> rewrite(Call expression, Captures captu
+            boolean[] clobInfo)
+    {
+        if (clobInfo[0] || clobInfo[1]) {
+            String firstArgument = firstExpression;
+            String secondArgument = secondExpression;
+            if (!clobInfo[0]) {
+                firstArgument = toNClob(firstExpression);
+            }
+            if (!clobInfo[1]) {
+                secondArgument = toNClob(secondExpression);
+            }
```
String firstArgument = clobInfo[0] ? firstExpression : toNClob(firstExpression);
String secondArgument = clobInfo[0] ? secondExpression : toNClob(secondExpression);

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteStringComparison.java:
```diff
@@ -70,19 +70,48 @@ public Optional<ParameterizedExpression> rewrite(Call expression, Captures captu
                                         .build())));
     }
 
+    private static String buildComparison(
```
separate commit

ssheikin replied:
or even PR.

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteStringComparison.java:
```diff
@@ -70,19 +70,48 @@ public Optional<ParameterizedExpression> rewrite(Call expression, Captures captu
 
-        if (isClob(firstArgument, context) || isClob(secondArgument, context)) {
-            return Optional.empty();
+        boolean[] clobInfo = {false, false};
+        if (isClob(firstArgument, context)) {
+            clobInfo[0] = true;
+        }
+        if (isClob(secondArgument, context)) {
+            clobInfo[1] = true;
```
can it be a part of buildComparison ?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+                }
+            }
+
+            for (CastTestCase testCase : unsupportedCastTypePushdown()) {
```
ProjectionPushdown may be more restrictive than Join criteria pushdown. Please test for joins separately.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+                .isFullyPushedDown();
+    }
+
+    private void assertJoinFullyPushedDown(Session session, String leftTable, String rightTable, String joinType, String joinCondition, List<Integer> expectedOutput)
```
consider to rename to `assertJoin` and return assertion itself.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+            Session session = joinPushdownEnabled(getSession());
+
+            // Implict cast between bounded varchar
+            assertJoinFullyPushedDown(session, leftTableName, rightTableName, "LEFT JOIN", "l.varchar_50 = r.varchar_100", asList(1, 2));
```
could you please asList -> List.of everywhere. So it looks the same and does not catch the attention.

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteStringComparison.java:
```diff
@@ -70,19 +70,48 @@ public Optional<ParameterizedExpression> rewrite(Call expression, Captures captu
+
+    private static String toNClob(String expression)
+    {
+        return "TO_NCLOB(%s)".formatted(expression);
```
What if during trino reads the remote DB the unsupported by trino type (some remote-specific type) was force-mapped to text, 
later here it was capturedAs CharType or VarcharType ?
Will TO_NCLOB work correctly? 
Maybe it's the case here too https://github.com/trinodb/trino/pull/22203/files#diff-510ce28e8ade87630227de284298df1cb44186f73bc9f2219e1f6ad6505d410aR66

krvikash replied:
removed this change. `nclob` cast pushdown is out of scope of this PR.

Praveen2112 commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,114 @@
+    private static final Capture<ConnectorExpression> VALUE = newCapture();
+    private final BiFunction<ConnectorSession, Type, WriteMapping> toWriteMapping;
+
+    public RewriteCast(BiFunction<ConnectorSession, Type, WriteMapping> toWriteMapping)
```
Can this function return the `targetType` as a `String` so that in future we could decouple the logic from `toWriteMapping` ?

Praveen2112 commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,114 @@
+    {
+        if (type instanceof CharType charType) {
+            if (charType.getLength() > ORACLE_CHAR_MAX_CHARS) {
+                return Optional.empty();
```
Can we mention the reason here ?

SemionPar replied:
Yes, something like `Oracle will throw an error on casts to char(n>ORACLE_CHAR_MAX_CHARS)`

Praveen2112 commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,114 @@
+    public Optional<JdbcExpression> rewrite(ConnectorExpression projectionExpression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        Call call = (Call) projectionExpression;
+        Type trinoType = call.getType();
```
How to `toTrinoType` ?

Praveen2112 commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+            }
+        }
+
+        private static List<CastTestCase> supportedCastTypePushdown()
```
From the rule it looked like `integer` to `varchar` or `char` looks like a supported cast pushdown - Can we restrict them or add test cases here ?

Praveen2112 commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteStringComparison.java:
```diff
@@ -70,19 +70,48 @@ public Optional<ParameterizedExpression> rewrite(Call expression, Captures captu
+            if (!clobInfo[1]) {
+                secondArgument = toNClob(secondExpression);
+            }
+            return "DBMS_LOB.compare(%s, %s) %s 0".formatted(firstArgument, secondArgument, operator);
```
Is the DBMS_LOB is something which is provided by default on Oracle or is it some sort of a package which needs to be installed ? How would it work with collation ? What if the underlying bytes in CLOB has some invalid string pattern - in this case won't Trino and Oracle differ ?

krvikash replied:
removed this change. nclob comparison is out of scope of this PR.

Praveen2112 commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +501,172 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+            }
+        }
+
+        private static List<CastTestCase> supportedCastTypePushdown()
```
This is same as parameterized tests ? Can we have an assertion which assert all types of join and we could pass the types as an argument to it ?

krvikash replied:
> This is same as parameterized tests ?

Yes
> Can we have an assertion which assert all types of join and we could pass the types as an argument to it ?

I have a test `testAllJoinPushdownWithCast` which covers all join but only with one data type. For other all data types I am verifying only with `INNER JOIN`. Should we test all data types with all join? Or one should be fine?

marcinsbd commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClient.java:
```diff
@@ -242,6 +245,12 @@ public OracleClient(
 
+        this.projectFunctionRewriter = new ProjectFunctionRewriter<>(
+                this.connectorExpressionRewriter,
+                ImmutableSet.<ProjectFunctionRule<JdbcExpression, ParameterizedExpression>>builder()
```
How about?
` ImmutableSet.of(new RewriteCast(this::toWriteMapping)))`

krvikash replied:
builder will reduce the difference when adding new rewrite class in future. So Ill keep the builder here.

krvikash commented:
Thanks @ssheikin @Praveen2112 @marcinsbd @SemionPar for the review. Addressed comments. Removing pushdown support for `nclob` data type. This requires some more work.

Here is summary:
This PR only adds support for the VARCHAR and CHAR cast pushdown, which has the following limitations:

1. Pushdown is not supported with Char having length > 500
2. Pushdown is not supported with varchar having length > 1000
3. Pushdown is not supported with unbounded varchar

krvikash commented:
Added `AbstractRewriteCast` in `base-jdbc` pacakge, so that other jdbc connector can support cast pushdown by implementing `toJdbcTypeHandle` and `buildCast`. Also I have the base test case in `base-jdbc` package.

There is test case enhancement required to cover this comment https://github.com/trinodb/trino/pull/22728#discussion_r1684151214 . I will create table using remote database instead of trino.

krvikash commented:
Now using remote databse executor to create the table which will cover all the type which is supported by trino.

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,107 @@
+    public CastDataTypeTest execute(SqlExecutor sqlExecutor, String tableNamePrefix)
+    {
+        checkState(!testCaseInputs.isEmpty(), "No test case inputs");
+        List<String> columnsWithType = new ArrayList<>();
+        for (TestCaseInput input : testCaseInputs) {
+            columnsWithType.add(input.columnName() + " " + input.columnType());
+        }
+        String tableDefinition = "(" + String.join(", ", columnsWithType) + ")";
```
streams api?

krvikash commented:
(Only some test data changed)

krvikash commented:
Thanks @SemionPar for the review. 

(Separated the logic for checking the eligibility for pushdown in a separate method and kept the `toJdbcTypeHandle` just to convert trino type to jdbc type)

krvikash commented:
(fixed typo)

krvikash commented:
(added test case with disable pushdown)

Praveen2112 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/AbstractRewriteCast.java:
```diff
@@ -0,0 +1,96 @@
+            return Optional.empty();
+        }
+
+        if (!eligibleForPushdown(sourceType, targetType)) {
+            return Optional.empty();
+        }
```
If we have this check first - then we don't have rewrite the expression for non supported types ?

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseCastPushdownTest.java:
```diff
@@ -0,0 +1,134 @@
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+
+@TestInstance(PER_CLASS)
+public abstract class BaseCastPushdownTest
```
Can we follow the testing guidelines as mentioned here - https://trino.io/docs/current/develop/tests.html

Praveen2112 replied:
Can it extend`AbstractTestQueryFramework` ? Since we it provides a lot of helper/util methods for assertion -

Praveen2112 replied:
If not it could be like a `SqlDataTypeTest`

krvikash replied:
As of now used `AbstractTestQueryFramework` till we decide to use `Nested`.

CC: @ssheikin

Praveen2112 reviewed (commented):
I need to review the test code once again

mayankvadariya commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    protected String buildCast(Type sourceType, Type targetType, String expression, String castType)
```
nit: should this method to have a default implementation?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseCastPushdownTest.java:
```diff
@@ -0,0 +1,134 @@
+
+    private void dropTable(CastDataTypeTest table)
+    {
+        if (table2 != null) {
```
what about table1 ?

krvikash replied:
Its typo :) It should be `table`

ssheikin reviewed (commented):
lgtm

krvikash commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,95 @@
+import static io.trino.plugin.oracle.OracleClient.ORACLE_CHAR_MAX_CHARS;
+import static io.trino.plugin.oracle.OracleClient.ORACLE_VARCHAR2_MAX_CHARS;
+
+public class RewriteCast
```
I think we should not mention varchar as in future we can support other types in same class.

krvikash commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    @Override
+    protected boolean eligibleForPushdown(Type sourceType, Type targetType)
```
Added test coverage and modified `eligibleForPushdown` to receive `JdbcTypeHandle` to handle this case.

krvikash commented:
Thanks @Praveen2112  @ssheikin @mayankvadariya for the review. Addressed comments.

Praveen2112 commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,99 @@
+                    OracleTypes.NCHAR,
+                    OracleTypes.NVARCHAR,
+                    OracleTypes.CLOB,
+                    OracleTypes.NCLOB -> true;
```
Since NCLOB doesn't support pushdown - does casting into varchar is being supported by Oracle - Does it break the existing semantics we have today ?

krvikash replied:
`nclob` as the source type can be type cast in `varchar` type. But otherway around is not possible directly using `CAST`.

```
SQL> select cast(c_nclob AS varchar2(10 char)) from presto_test_user.table_1_dn5kb0yskx;

CAST(C_NCL
----------
India
Poland


SQL> select cast(C_VARCHAR_10 AS nclob) from presto_test_user.table_1_dn5kb0yskx;
select cast(C_VARCHAR_10 AS nclob) from presto_test_user.table_1_dn5kb0yskx
       *
ERROR at line 1:
ORA-00932: inconsistent datatypes: expected - got NCLOB
```

Praveen2112 replied:
NCLOB does case insensitive search right ? So in that case can we have a test case wrt to join which ensures join condition is not applied in a case insensitive way

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public class CastDataTypeTest
```
Should it be final ?

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+        {
+            requireNonNull(columnName, "columnName is null");
+            requireNonNull(columnType, "columnType is null");
+            requireNonNull(inputValues, "inputValues is null");
```
ImmutableList#copyOf

krvikash replied:
elements could be null. So Ill use `List#copyOf` here

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+            requireNonNull(targetColumn, "targetColumn is null");
+        }
+
+        public CastTestCase(String sourceColumn, String castType)
+        {
+            this(sourceColumn, castType, Optional.empty());
+        }
```
Can we have it before the record ctor ?

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+        if (inputSize.isPresent()) {
+            List<String> rowToInsert = new ArrayList<>();
+            for (int i = 0; i < inputSize.getAsInt(); i++) {
+                List<String> row = new ArrayList<>();
```
Can we use a StringJoiner to build the String and pass it as a row ?

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+                .collect(joining(", ", "(", ")"));
+
+        if (inputSize.isPresent()) {
+            List<String> rowToInsert = new ArrayList<>();
```
nit: ImmutableList#builder

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,109 @@
+    private final List<TestCaseInput> testCaseInputs = new ArrayList<>();
+
+    private OptionalInt inputSize = OptionalInt.empty();
+    private TestTable testTable;
```
Where do we drop this `TestTable` ?

Praveen2112 commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +497,38 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+            Session session = joinPushdownEnabled(getSession());
+
+            // Implict cast between bounded varchar
+            assertJoin(session, leftTableName, rightTableName, "LEFT JOIN", "l.varchar_50 = r.varchar_100")
```
@ssheikin Would have a different opinion - But I would like to see the query being inlined here

ssheikin replied:
In this case inlining of method is justified. ..but `"SELECT id FROM %s l %s %s r ON %s"` may be extracted as variable 😄

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,119 @@
+public abstract class BaseJdbcCastPushdownTest
+        extends AbstractTestQueryFramework
+{
+    protected CastDataTypeTest table1;
+    protected CastDataTypeTest table2;
```
These table are used for join right ? Can we name it as `left` and `right` ?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOracleCastPushdown.java:
```diff
@@ -0,0 +1,69 @@
+    @AfterAll
+    public final void destroy()
+    {
+        oracleServer = null; // closed by closeAfterClass
```
not needed

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOracleCastPushdown.java:
```diff
@@ -0,0 +1,69 @@
+
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+
+@TestInstance(PER_CLASS)
```
CONCURRENT?

krvikash replied:
I did not see any test with `@TestInstance(CONCURRENT)`

krvikash commented:
Thanks for review @Praveen2112 @ssheikin. Addressed comments.

krvikash commented:
(removed some unused oracle types)

Praveen2112 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/AbstractRewriteCast.java:
```diff
@@ -0,0 +1,100 @@
+            return Optional.empty();
+        }
+
+        Optional<JdbcTypeHandle> targetTypeJdbcHandle = toJdbcTypeHandle(targetType);
+        if (targetTypeJdbcHandle.isEmpty()) {
+            return Optional.empty();
+        }
```
Can we have it before rewriting the expression so that we could skip rewriting if the target type is not supported.

Praveen2112 commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/AbstractRewriteCast.java:
```diff
@@ -0,0 +1,100 @@
+
+    protected abstract boolean pushdownSupported(JdbcTypeHandle sourceType, Type targetType);
+
+    protected abstract Optional<JdbcTypeHandle> toJdbcTypeHandle(Type type);
```
In order to identify a few additional information like case sensitivity - Do we need pass on the JdbcTypeHandle also - and what if we could merge it with `pushdownSupported` since the API looks same.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleCastPushdownTest.java:
```diff
@@ -0,0 +1,382 @@
+    public void testCastPushdownSpecialCase()
+    {
+        for (CastTestCase testCase : specialCaseNClob()) {
+            // Projection pushdown is supported, because The clob type internally converted to nclob thus cast does not happen
```
> internally

In oracle?

krvikash replied:
No, In trino

ssheikin replied:
please improve comment

krvikash replied:
Modifying comment to
```
// Projection pushdown is supported, because trino converts the clob type to nclob thus cast is not required
```

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    protected void assertJoinFullyPushedDown(String leftTable, String rightTable, String joinType, String joinCondition)
```
inline? 
Looks like inlined it will be read the same as extracted to a method

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,116 @@
+                .isFullyPushedDown();
+    }
+
+    protected void assertJoinNotFullyPushedDown(String leftTable, String rightTable, String joinCondition)
```
inline?
Looks like inlined it will be read the same as extracted to a method

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,116 @@
+
+    protected void assertJoinNotFullyPushedDown(String leftTable, String rightTable, String joinCondition)
+    {
+        assertJoin(leftTable, rightTable, "JOIN", joinCondition)
```
why `assertJoinFullyPushedDown` has `joinType` as parameter, and here it's hardcoded?

krvikash replied:
This method will be removed after inlining.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +496,35 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+            Session session = joinPushdownEnabled(getSession());
+
+            // Implicit cast between bounded varchar
+            String joinWithBoundedVarchar = "SELECT id FROM " + leftTableName + " l %s " + rightTableName + " r ON l.varchar_50 = r.varchar_100";
```
.formatted instead of concat?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +496,35 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+                    .isFullyPushedDown();
+
+            // Implicit cast between bounded and unbounded varchar
+            String joinWithUnboundedVarchar = "SELECT id FROM " + leftTableName + " l %s " + rightTableName + " r ON l.varchar_50 = r.varchar_unbounded";
```
.formatted instead of concat?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+        return testTable.getName();
+    }
+
+    public void close()
```
AutoClosable?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+        return this;
+    }
+
+    public String tableName()
```
toString()?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+        {
+            requireNonNull(columnName, "columnName is null");
+            requireNonNull(columnType, "columnType is null");
+            inputValues = new ArrayList<>(requireNonNull(inputValues, "inputValues is null"));
```
ImmutableList.copyOf

krvikash replied:
`inputValues` can have `null` elements. So can't use `ImmutableList.copyOf` here.

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+        String tableDefinition = columnsWithType.stream()
+                .collect(joining(", ", "(", ")"));
+
+        ImmutableList.Builder<String> rowToInsert = ImmutableList.builder();
```
rowToInsert ->
rowsToInsert

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+
+    public CastDataTypeTest addColumn(String columnName, String columnType, List<Object> inputValues)
+    {
+        if (inputSize.isEmpty()) {
+            inputSize = OptionalInt.of(inputValues.size());
+        }
+        checkArgument(inputSize.getAsInt() == inputValues.size(), "Expected input size: %s, but found: %s", inputSize.getAsInt(), inputValues.size());
```
why is it needed?

krvikash replied:
Just to ensure that caller should not provide the mismatched length for `inputValues`

ssheikin replied:
why caller should not provide the mismatched length for inputValues ?

krvikash replied:
`inputValues`'s size indicates the total number of rows to be inserted into the table. If we have mismatched size for other columns, then It won't be valid.

I think I will `public static CastDataTypeTest create()` can be changed to `public static CastDataTypeTest create(int rowSize)` to accept the row size.

ssheikin replied:
I think it's better to haverow size implicit. 

Maybe `Suppliers.memoize` ?

ssheikin replied:
`Suppliers.memoize` will not verify. so no CR here.

ssheikin replied:
I liked an implicit approach, but explicit looks good too.

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,113 @@
+        ImmutableList.Builder<String> rowToInsert = ImmutableList.builder();
+        if (inputSize.isPresent()) {
+            for (int i = 0; i < inputSize.getAsInt(); i++) {
+                StringJoiner rowBuilder = new StringJoiner(",");
```
why here `StringJoiner` and above `collect(joining(` ?

krvikash replied:
It was suggested by Praveen https://github.com/trinodb/trino/pull/22728#discussion_r1709338703

ssheikin replied:
```
        ImmutableList<String> rowsToInsert = IntStream.range(0, inputSize.orElse(0))
                .mapToObj(rowId -> testCaseInputs.stream()
                        .map(TestCaseInput::inputValues)
                        .map(rows -> rows.get(rowId))
                        .map(String::valueOf)
                        .collect(joining(",")))
                .collect(toImmutableList());
```

marcinsbd commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,116 @@
+    {
+        // combination of supported types
+        for (CastTestCase testCase : supportedCastTypePushdown()) {
+            assertJoinFullyPushedDown(left.tableName(), right.tableName(), "JOIN", "CAST(l.%s AS %s) = r.%s".formatted(testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow()));
```
nit: should we wrap this line ( line per argument)? how do you feel about it?

krvikash replied:
I think It will be easy to read in one line rather than multiple lines since we have multiple assert of these kinds.

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleCastPushdownTest.java:
```diff
@@ -0,0 +1,382 @@
+                new CastTestCase("c_number_10", "varchar(50)", Optional.of("c_varchar_50")),
+                new CastTestCase("c_number_19", "varchar(50)", Optional.of("c_varchar_50")),
+                new CastTestCase("c_number_15", "varchar(50)", Optional.of("c_varchar_50")),
+                new CastTestCase("c_number_10_2", "varchar(50)", Optional.of("c_varchar_50")),
```
In test methods except of pushdown behaviour we are interested in asserting actual result too.

krvikash replied:
Won't the test will be crowded? I think asserting pushdown also verify the result as well.

ssheikin replied:
Right, `Verifies query is fully pushed down and that results are the same as when pushdown is fully disabled.`

krvikash commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,93 @@
+    private static boolean supportedSourceTypeToCastInVarchar(JdbcTypeHandle sourceType)
+    {
+        return switch (sourceType.jdbcType()) {
+            case OracleTypes.NUMBER,
```
No, The tests are added here `testCastFails`

ssheikin replied:
even `1` -> `"1"` ?

krvikash replied:
Yes,

![image](https://github.com/user-attachments/assets/f9492059-5815-4ade-9eb9-0eac85807a45)

![image](https://github.com/user-attachments/assets/ffecf4bd-4676-4faf-84fe-6ac3087469a6)

krvikash commented:
Thanks @ssheikin @Praveen2112 @marcinsbd for the review. Addressed comments.

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,112 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public final class CastDataTypeTest
```
IIRC previously it was named `...Table`, wasn't it?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,112 @@
+public final class CastDataTypeTest
+        implements AutoCloseable
+{
+    private final List<TestCaseInput> testCaseInputs = new ArrayList<>();
```
rows?

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,112 @@
+import static java.util.stream.Collectors.joining;
+
+public final class CastDataTypeTest
+        implements AutoCloseable
```
it may be even TemporaryRelation

ssheikin commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/OracleClient.java:
```diff
@@ -793,6 +808,11 @@ private SliceWriteFunction oracleCharWriteFunction()
         return SliceWriteFunction.of(Types.NCHAR, (statement, index, value) -> statement.unwrap(OraclePreparedStatement.class).setFixedCHAR(index, value.toStringUtf8()));
     }
 
+    private String toTargetType(ConnectorSession session, Type type)
```
Not sure if `toTargetType` fits to `...Client` class.
Why not to inline this method?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:
```diff
@@ -496,6 +496,35 @@ private void predicatePushdownTest(String oracleType, String oracleLiteral, Stri
+                    .isFullyPushedDown();
+            assertThat(query(session, joinWithBoundedVarchar.formatted("INNER JOIN")))
+                    .isFullyPushedDown();
+
```
FULL, CROSS JOIN?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleCastPushdownTest.java:
```diff
@@ -0,0 +1,431 @@
+                .isFullyPushedDown();
+        assertJoin(leftTable(), rightTable(), "INNER JOIN", "CAST(l.%1$s AS %3$s) = CAST(r.%2$s AS %3$s)".formatted(testCase.sourceColumn(), testCase.targetColumn().orElseThrow(), testCase.castType()))
+                .isFullyPushedDown();
+    }
```
FULL, CROSS JOIN?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOracleCastPushdown.java:
```diff
@@ -0,0 +1,65 @@
+        oracleServer = closeAfterClass(new TestingOracleServer());
+        return OracleQueryRunner.builder(oracleServer)
+                .addConnectorProperties(ImmutableMap.<String, String>builder()
+                        .put("oracle.connection-pool.enabled", "false")
```
does it relate to test?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/TestOracleCastPushdown.java:
```diff
@@ -0,0 +1,65 @@
+        return OracleQueryRunner.builder(oracleServer)
+                .addConnectorProperties(ImmutableMap.<String, String>builder()
+                        .put("oracle.connection-pool.enabled", "false")
+                        .put("oracle.remarks-reporting.enabled", "true")
```
does it relate to test?

krvikash commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTest.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public void close()
+    {
+        testTable.close();
```
`closeAllRuntimeException` is for `Closable` object. `testTable` is `AutoClosable`. 

Is it fine to check null and then close?
```
        if (testTable != null) {
            testTable.close();
        }
```

ssheikin replied:
Yes, it's fine.
Also, `public interface Closeable extends AutoCloseable`

ssheikin replied:
You can also use try with resources.

ssheikin replied:
Or even `io.airlift.testing.Closeables#closeAll(java.lang.AutoCloseable...)`

krvikash commented:
Thanks @ssheikin, AC.

krvikash commented:
(some code refactoring)

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,100 @@
+
+    protected QueryAssertions.QueryAssert assertJoin(String leftTable, String rightTable, String joinType, String joinCondition)
+    {
+        return assertThat(query("SELECT l.id FROM %s l %s %s r ON %s".formatted(leftTable, joinType, rightTable, joinCondition)));
```
do you want to inline method, extracting variable for query?
Like `testJoinPushdownWithImplicitCast`

ssheikin commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/CastDataTypeTestTable.java:
```diff
@@ -0,0 +1,103 @@
+        try {
+            closeAll(testTable);
+        }
+        catch (Exception e) {
```
make method `throws`

krvikash replied:
base class `close` method does not throws. So can't make this method `throws`. May be as a followup?

ssheikin commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleCastPushdownTest.java:
```diff
@@ -0,0 +1,436 @@
+    private CastDataTypeTestTable right;
+
+    @BeforeAll
+    public void setupTable()
```
setupTable ->
setupTables

or something even more generic and simple as `init`

krvikash replied:
`init` is already part of `AbstractTestQueryFramework`. Ill name it `setup` then.

mosabua commented:
Adding stale-ignore label since this looks like its very active but just stalled a bit.

krvikash commented:
(Do not cast unnecessary with extra space padding when target char type has more length than source char type)

krvikash commented:
(rebased with master)

krvikash commented:
(CI fix)

Praveen2112 commented on plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleCastPushdownTest.java:
```diff
@@ -0,0 +1,436 @@
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public abstract class BaseOracleCastPushdownTest
```
Do we need this abstraction ?

krvikash replied:
Abstraction will be useful in the scenario where we might want to verify cast pushdown for different query runner.

krvikash commented on plugin/trino-oracle/src/main/java/io/trino/plugin/oracle/RewriteCast.java:
```diff
@@ -0,0 +1,108 @@
+        if (sourceType instanceof CharType sourceCharType && targetType instanceof CharType targetCharType) {
+            if (sourceCharType.getLength() < targetCharType.getLength()) {
+                // Do not cast unnecessary with extra space padding when target char type has more length than source char type
+                return expression;
```
Here are the execution plans. When we skip rewriting `ScanProject` node gets populated and when we perform pushdown `TableScan` node is populated. So In terms of the test case, we can not use `isFullyPushedDown` when we skip rewriting.

When we skip rewriting,

```
MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [expr]
    Output partitioning: SINGLE []
    Output[columnNames = [_col0]]
    │   Layout: [expr:char(32)]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    │   _col0 := expr
    └─ ScanProject[table = oracle:presto_test_user.left_table_se623gecjf PRESTO_TEST_USER.LEFT_TABLE_SE623GECJF columns=[C_CHAR_10:char(10):CHAR]]
           Layout: [expr:char(32)]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}/{rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           expr := CAST(c_char AS char(32))
           c_char := C_CHAR_10:char(10):CHAR

]], types=[varchar(650)], setSessionProperties={}, resetSessionProperties=[]}
```

And we perform pushdown,

```
MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [pfgnrtd]
    Output partitioning: SINGLE []
    Output[columnNames = [_col0]]
    │   Layout: [pfgnrtd:char(32)]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    │   _col0 := pfgnrtd
    └─ TableScan[table = oracle:Query[SELECT "C_CHAR_10" AS "pfgnrtd_0" FROM "PRESTO_TEST_USER"."LEFT_TABLE_BJL0LGPNFS"] columns=[pfgnrtd_0:char(32):char]]
           Layout: [pfgnrtd:char(32)]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           pfgnrtd := pfgnrtd_0:char(32):char

]], types=[varchar(581)], setSessionProperties={}, resetSessionProperties=[]}
```

krvikash replied:
rewriting has a advantage that we could use `isFullyPushedDown` to verify the pushdown in such cases.

krvikash replied:
Also not rewriting cast for this will result in `JOIN` not pushdown for `SELECT l.id FROM table1 l JOIN table2 r ON CAST(l.c_char_10 AS char(50)) = r.c_char_50;`.

```
Output[columnNames = [id]]
│   Layout: [id:decimal(10,0)]
└─ InnerJoin[criteria = (expr = c_char_26), distribution = PARTITIONED]
   │   Layout: [id:decimal(10,0)]
   │   Distribution: PARTITIONED
   ├─ RemoteExchange[type = REPARTITION]
   │  │   Layout: [id:decimal(10,0), expr:char(50)]
   │  └─ ScanProject[table = oracle:presto_test_user.left_table_wf7cu8lky4 PRESTO_TEST_USER.LEFT_TABLE_WF7CU8LKY4 columns=[ID:decimal(10,0):NUMBER, C_CHAR_10:char(10):CHAR]]
   │         Layout: [id:decimal(10,0), expr:char(50)]
   │         expr := CAST(c_char AS char(50))
   │         c_char := C_CHAR_10:char(10):CHAR
   │         id := ID:decimal(10,0):NUMBER
   └─ LocalExchange[partitioning = HASH, arguments = [c_char_26::char(50)]]
      │   Layout: [c_char_26:char(50)]
      └─ RemoteExchange[type = REPARTITION]
         │   Layout: [c_char_26:char(50)]
         └─ TableScan[table = oracle:presto_test_user.right_table_g1rzz51027 PRESTO_TEST_USER.RIGHT_TABLE_G1RZZ51027 columns=[C_CHAR_50:char(50):CHAR]]
                Layout: [c_char_26:char(50)]
                c_char_26 := C_CHAR_50:char(50):CHAR
```

Praveen2112 commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,94 @@
+    public void testCastFails()
+    {
+        for (CastTestCase testCase : failCast()) {
+            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                    .hasMessageMatching("(.*)Cannot cast (.*) to (.*)");
```
Does it thrown by Trino or by Oracle or by both ?

krvikash replied:
By Trino.

krvikash commented:
Pushdown was giving incorrect result wrt Unicode character values and `char` data type column. So, For now in this PR I am removing `char` cast pushdown.

krvikash commented:
(rebased with master)

krvikash commented:
> Pushdown was giving incorrect result wrt Unicode character values and `char` data type column. So, For now in this PR I am removing `char` cast pushdown.

The issue seems to be happening in this case: When casting `char` unicode value to `varchar` type.

```
SELECT CAST(c_char_unicode AS varchar(50)) FROM table1;
```
```
SELECT CAST(c_nchar_unicode AS varchar(50)) FROM table1;
```

mayankvadariya commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/AbstractRewriteCast.java:
suggestion: Support cast projection pushdown in oracle -> Support char/varchar cast projection pushdown in oracle

Praveen2112 replied:
Support varchar cast pushdown in Oracle

krvikash replied:
@Praveen2112 we are supporting char pushdown as well with some limitation.

krvikash replied:
https://github.com/trinodb/trino/pull/22728#issuecomment-2357521017

mayankvadariya reviewed (approved):
Please evaluate a possibility of case-insensitive matching in Oracle when pushed down.

SemionPar commented on plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java:
```diff
@@ -0,0 +1,94 @@
+    public void testCastFails()
+    {
+        for (CastTestCase testCase : failCast()) {
+            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
```
When this test fails, it doesn't say which condition has failed exactly. Do you think we could improve on that?
Took a quick look - `as()` or `withFailMessage()` does not seem to work, maybe custom `satifies()` would.

krvikash replied:
Thanks @SemionPar, I will take this as a follow up.

SemionPar replied:
@krvikash Sure thing!

One more thought that crossed my mind: if cast fails at runtime with different message when pushed down, say:

```
      Trino: Cannot cast DECIMAL(19, 2) '99999999999999999.99' to DECIMAL(10, 2)
      Teradata: [Error 2616] [SQLState 22003] Numeric overflow occurred during computation.
```

It would be good to be able to use `testCastFails` to test such cases, don't you think? We might want to add message customization to `CastTestCase` (or introduce `FailCastTestCase`) - consider this another improvement idea.

krvikash replied:
@SemionPar Here is the refactoring PR https://github.com/trinodb/trino/pull/23737

ssheikin commented:
🎉
